src/main/kotlin/com/lykke/trade/volume/monitoring/service/process/impl/EventDeduplicationServiceImpl.kt

```diff
@@ -1,16 +1,64 @@
 package com.lykke.trade.volume.monitoring.service.process.impl
 
-import com.lykke.trade.volume.monitoring.service.entity.ProcessedEvent
+import com.lykke.trade.volume.monitoring.service.exception.ApplicationException
+import com.lykke.trade.volume.monitoring.service.loader.EventsLoader
 import com.lykke.trade.volume.monitoring.service.process.EventDeduplicationService
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.TaskScheduler
+import java.time.Duration
+import java.time.ZonedDateTime
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
 
-// todo: additional task: implement deduplication
-class EventDeduplicationServiceImpl : EventDeduplicationService {
-    override fun isDuplicate(sequenceNumber: Long): Boolean {
-        return false
+class EventDeduplicationServiceImpl(private val eventsLoader: EventsLoader,
+                                    private val cleanPeriod: Long,
+                                    private val taskScheduler: TaskScheduler) : EventDeduplicationService {
+
+    companion object {
+        private val LOGGER = LoggerFactory.getLogger(EventDeduplicationServiceImpl::class.java.name)
+    }
+
+    private val lock = ReentrantLock()
+    private var initialized = false
+    private var sequenceNumbers: MutableSet<Long> = HashSet()
+    private var prevSequenceNumbers: Set<Long>? = null
+
+    @Synchronized
```

Comment:
suggest to use only one lock here `lock` as lock inside the lock is deadlock prone

Reply:
Decided to leave it as it is as these locks have different purposes